### PR TITLE
cli: (program/buffer deploy) introduce --fee-payer argument

### PR DESCRIPTION
**Pull request by norwnd**
#### Problem

`solana program deploy` and `solana program write-buffer` could use a `--fee-payer` argument.

#### Summary of Changes

This adds `--fee-payer` to `solana program deploy` and `solana program write-buffer` cli commands. It's effectively a spin-off from https://github.com/solana-labs/solana/pull/33860.

### CODE REVIEW

File: cli/src/program.rs [hunk 1]
```diff
@@ -1027,21 +1053,28 @@ fn process_program_deploy(
...
         program_len
     } else {
         program_len * 2
     };
-    let minimum_balance = rpc_client.get_minimum_balance_for_rent_exemption(
-        UpgradeableLoaderState::size_of_programdata(programdata_len),
-    )?;
 
-    let result = if do_deploy {
+    let (min_rent_exempt_program_balance, min_rent_exempt_program_data_balance) = (
```
Comment (joncinque) [inline cli/src/program.rs:1069]: Sorry, why is this change for the rent-exemption of the base program included? In general, please avoid putting in unrelated changes, it makes each review much more difficult.

File: cli/src/program.rs [hunk 2]
```diff
@@ -2076,23 +2123,20 @@ fn do_process_program_write_and_deploy(
...
 }
 
+#[allow(clippy::too_many_arguments)]
 fn do_process_program_upgrade(
     rpc_client: Arc<RpcClient>,
     config: &CliConfig,
-    program_data: &[u8],
-    program_id: &Pubkey,
+    program_data: &[u8], // can be empty, hence we have program_len
+    program_len: usize,
```
Comment (joncinque) [inline cli/src/program.rs:2131]: While I appreciate the fix here, it's not clear at all why this change was made, so please explain when you're making unrelated changes in the PR description, or put the change in a separate commit, or even better, make a separate PR!

Just so I'm sure I understand, the current behavior of reusing an existing buffer for a deployment will always succeed the check in  `complete_partial_program_init` of:

``` 
if !account.data.is_empty() && account.data.len() < account_data_len {
```

because `account_data_len` will be incorrectly zero, is that it?

File: cli/src/program.rs [hunk 3]
```diff
@@ -1925,11 +1971,13 @@ where
 fn do_process_program_write_and_deploy(
     rpc_client: Arc<RpcClient>,
     config: &CliConfig,
-    program_data: &[u8],
+    program_data: &[u8], // can be empty, hence we have program_len
     program_len: usize,
-    programdata_len: usize,
-    minimum_balance: u64,
+    min_rent_exempt_program_balance: u64,
```
Comment (joncinque) [inline cli/src/program.rs:1976]: Can you revert this change? There are some tricky code paths around here when it comes to non-upgradeable deployments, and this creates more footguns.

In the case of a buffer that already exists, they need to pass the program's rent-exemption in `min_rent_exempt_program_data_balance`, but in the case of a new deployment, they would need to pass that value in `min_rent_exempt_program_balance`, which is very mistake-prone.

**Review by joncinque — COMMENTED** [review]
Looks good for the most part!

As a reminder, please avoid putting in unrelated changes, it makes the review much more complicated for me. These code paths have little gotchas all over the place, so changing a parameter can have unintended consequences, so I have to trace every single possible path to make sure it's ok.

File: cli/src/program.rs (hunk 1 above)
Comment (norwnd) [reply cli/src/program.rs:1069]: > Sorry, why is this change for the rent-exemption of the base program included?

I believe this rewrite isn't changing any functionality, just a refactoring.

The original motivation was the need to accommodate `--sign-only` for `deploy`, it's not really necessary now (after we decided to ditch offline signing for `deploy`), but I was also hoping to make it a bit more readable (by grouping related stuff together) so I don't have to "untangle it" again next time I have to go through that code.

For reviewer(s) it's not as pleasant of a thing ... I get that, trying to find the right balance between those ^

File: cli/src/program.rs (hunk 3 above)
Comment (norwnd) [reply cli/src/program.rs:1976]: Not sure how this change makes it more of a foot-gun, if anything I was hoping to remove a foot-gun or two with it - by making it explicit what `fn do_process_program_write_and_deploy` actually expects as inputs,

it's not strictly needed anymore (it was for enabling `--sign-only` in `deploy`) but it kinda makes me sad to discover this code is "that rigid", unless it's not expected to change much in the future I'd say it's a sign that refactoring(s) are needed to untangle it a bit.

So, do you want me to revert the parts touching "balances" ? Or something else ?

File: cli/src/program.rs [hunk 4]
```diff
@@ -2076,23 +2123,20 @@ fn do_process_program_write_and_deploy(
...
 fn do_process_program_upgrade(
     rpc_client: Arc<RpcClient>,
     config: &CliConfig,
-    program_data: &[u8],
-    program_id: &Pubkey,
+    program_data: &[u8], // can be empty, hence we have program_len
+    program_len: usize,
+    min_rent_exempt_program_data_balance: u64,
+    fee_payer_signer: &dyn Signer,
+    program_pubkey: &Pubkey,
```
[the remark this answers is not in the document]
Comment (norwnd) [reply cli/src/program.rs]: Applied in https://github.com/solana-labs/solana/pull/34341/commits/a2a84f8ffeb479376f3b9fd27f0c8bd4ad7b166c

File: cli/src/program.rs (hunk 2 above)
Comment (joncinque) [reply cli/src/program.rs:2131]: Ok great, thanks!

File: cli/src/program.rs (hunk 1 above)
Comment (joncinque) [reply cli/src/program.rs:1069]: > PS: when lots of small changes touch same parts of the code it's typically more efficient and safe (no merge conflicts to resolve, can test the final result just once while aware of all the relevant context) to change it all in one fell swoop, that's why it's all here.

While it's definitely more efficient for the developer to just make all the changes at once, it's much less efficient for the maintainers because they need to deal with a refactor *and* a code change at the same time.

Those kinds of pull requests are often the most dangerous. If you look back at the bugs introduced in codebases, it's almost always because of a refactor that also changed behavior. Because they're often big, the reviewer doesn't look at it so closely, and assumes that tests will catch any issues.

While you can work however you want, when it comes to these repos, if you do a refactor, you will be asked to do it in a separate PR.

**Review by joncinque — APPROVED** [review]
Great work, thanks so much!
